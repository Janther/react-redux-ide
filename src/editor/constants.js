export const EDITOR_FOCUSED = "EDITOR_FOCUSED";
