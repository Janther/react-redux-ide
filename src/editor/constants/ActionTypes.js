export const DATA_FETCHED = 'DATA_FETCHED';
export const DATA_SUCCEEDED = 'DATA_SUCCEEDED';
export const DATA_FAILED = 'DATA_FAILED';

export const EDITOR_TEXT_CHANGED = 'TEXT_CHANGED';
export const EDITOR_MOVE_CURSOR  = 'MOVE_CURSOR';

export const EDITOR_UPDATE_CHAR_SIZE = 'UPDATE_CHAR_SIZE';
export const EDITOR_INVALIDATE_CHAR_SIZE  = 'INVALIDATE_CHAR_SIZE';
