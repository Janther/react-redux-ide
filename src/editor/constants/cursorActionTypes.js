export const EDITOR_MOVE_CURSOR  = 'MOVE_CURSOR';
