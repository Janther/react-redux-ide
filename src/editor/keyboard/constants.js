export const EDITOR_LINE_CHANGED = "LINE_CHANGED";
export const EDITOR_NEW_LINE = "EDITOR_NEW_LINE";
export const EDITOR_MOVE_UP_CURSOR = "MOVE_UP_CURSOR";
export const EDITOR_MOVE_DOWN_CURSOR = "MOVE_DOWN_CURSOR";
export const EDITOR_MOVE_LEFT_CURSOR = "MOVE_LEFT_CURSOR";
export const EDITOR_MOVE_RIGHT_CURSOR = "MOVE_RIGHT_CURSOR";
export const EDITOR_BACKSPACE = "BACKSPACE";
export const EDITOR_DEL = "DEL";
export const SUBSCRIBE_KEYBOARD = "SUBSCRIBE_KEYBOARD";
export const EDITOR_SELECT_UP = "EDITOR_SELECT_UP";
export const EDITOR_SELECT_DOWN = "EDITOR_SELECT_DOWN";
export const EDITOR_SELECT_LEFT = "EDITOR_SELECT_LEFT";
export const EDITOR_SELECT_RIGHT = "EDITOR_SELECT_RIGHT";
