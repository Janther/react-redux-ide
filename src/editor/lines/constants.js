export const EDITOR_UPDATE_CHAR_SIZE = "UPDATE_CHAR_SIZE";
export const EDITOR_INVALIDATE_CHAR_SIZE = "INVALIDATE_CHAR_SIZE";
