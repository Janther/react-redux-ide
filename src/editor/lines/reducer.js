import * as constants from "./constants";

const lines = function(
  state = {
    invalidCharSize: true,
    charSize: {
      defaultCharWidth: 0,
      doubleWidthCharWidth: 0,
      halfWidthCharWidth: 0,
      koreanCharWidth: 0,
      lineHeightInPixels: 0
    }
  },
  action
) {
  switch (action.type) {
    case constants.EDITOR_UPDATE_CHAR_SIZE:
      return {
        ...state,
        invalidCharSize: false,
        charSize: action.size
      };
    case constants.EDITOR_INVALIDATE_CHAR_SIZE:
      return {
        ...state,
        invalidCharSize: true
      };
    default:
      return state;
  }
};

export default lines;
