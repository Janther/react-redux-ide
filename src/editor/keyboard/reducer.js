import { combineReducers } from "redux";
import textarea from "./reducers/textarea";
import commands from "./reducers/commands";
import keyboardWrapper from "./reducers/keyboardWrapper";
import lines from "./reducers/lines";
import cursor from "./reducers/cursor";

const keyboard = combineReducers({
  keyboardWrapper,
  textarea,
  commands,
  lines,
  cursor
});

export default keyboard;
