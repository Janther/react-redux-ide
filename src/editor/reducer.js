import { combineReducers } from 'redux';
import { createReducer } from './utils/reducerUtils';
import * as constants from './constants';
import keyboard from './keyboard/reducer';
import lines from './lines/reducer';
import fromPairs from 'lodash.frompairs';

const focusEditor = function(state, action) ({
  ...state,
  selectedEditor: action.selectedEditor
});

const editors = createReducer(
  {
    editors: ['style'],
    selectedEditor: 'style'
  },
  fromPairs([
    [constants.EDITOR_FOCUSED, focusEditor]
  ])
);

const editorReducer = combineReducers({
  keyboard,
  lines,
  editors
});

export default editorReducer;
