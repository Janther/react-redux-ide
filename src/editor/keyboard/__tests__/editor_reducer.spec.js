import reducer from '../reducers/cursor'
import * as constants from '../constants'

describe('cursor reducer', () => {
  it('should return the initial state', () => {
    expect(
      reducer(undefined, {})
    ).toMatchSnapshot();
  });

  it('should handle EDITOR_MOVE_CURSOR up', () => {
    expect(
      reducer({ lineIndex: 0, charIndex: 0 }, {
        type: constants.EDITOR_MOVE_UP_CURSOR,
        payload: {
          lines: []
        }
      })
    ).toMatchSnapshot();

    expect(
      reducer({ lineIndex: 0, charIndex: 0 }, {
        type: constants.EDITOR_MOVE_UP_CURSOR,
        payload: {
          lines: [{
            value: '1'
          }]
        }
      })
    ).toMatchSnapshot();

    expect(
      reducer({ lineIndex: 0, charIndex: 1 }, {
        type: constants.EDITOR_MOVE_UP_CURSOR,
        payload: {
          lines: [{
            value: '1'
          }]
        }
      })
    ).toMatchSnapshot();

    expect(
      reducer({ lineIndex: 2, charIndex: 0 }, {
        type: constants.EDITOR_MOVE_UP_CURSOR,
        payload: {
          lines: [{
            value: '1'
          }, {
            value: '2',
          }, {
            value: '3'
          }]
        }
      })
    ).toMatchSnapshot();

    expect(
      reducer({ lineIndex: 2, charIndex: 1 }, {
        type: constants.EDITOR_MOVE_UP_CURSOR,
        payload: {
          lines: [{
            value: '1'
          }, {
            value: '2',
          }, {
            value: '3'
          }]
        }
      })
    ).toMatchSnapshot();
  });

  it('should handle EDITOR_MOVE_CURSOR down', () => {
    expect(
      reducer({ lineIndex: 1, charIndex: 0 }, {
        type: constants.EDITOR_MOVE_DOWN_CURSOR,
        payload: {
          lines: [{
            value: '1'
          }, {
            value: '2',
          }]
        }
      })
    ).toMatchSnapshot();

    expect(
      reducer({ lineIndex: 0, charIndex: 0 }, {
        type: constants.EDITOR_MOVE_DOWN_CURSOR,
        payload: {
          lines: [{
            value: '1'
          }, {
            value: '2',
          }]
        }
      })
    ).toMatchSnapshot();
  });

  it('should handle EDITOR_MOVE_CURSOR left', () => {
    expect(
      reducer({ lineIndex: 0, charIndex: 0 }, {
        type: constants.EDITOR_MOVE_CURSOR,
        payload: {
          lines: [{
            value: '1'
          }, {
            value: '2',
          }]
        }
      })
    ).toMatchSnapshot();

    expect(
      reducer({ lineIndex: 0, charIndex: 2 }, {
        type: constants.EDITOR_MOVE_CURSOR,
        payload: {
          lines: [{
            value: '1'
          }, {
            value: '2',
          }]
        }
      })
    ).toMatchSnapshot();
  });

  it('should handle EDITOR_MOVE_CURSOR right', () => {
    expect(
      reducer({ lineIndex: 0, charIndex: 0 }, {
        type: constants.EDITOR_MOVE_CURSOR,
        payload: {
          lines: [{
            value: '1'
          }, {
            value: '2',
          }]
        }
      })
    ).toMatchSnapshot();
  });
})
