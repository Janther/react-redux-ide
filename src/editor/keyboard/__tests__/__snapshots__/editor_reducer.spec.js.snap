exports[`cursor reducer should handle EDITOR_MOVE_CURSOR down 1`] = `
Object {
  "charIndex": 1,
  "endOffset": 1,
  "lineIndex": 1,
  "startOffset": 1,
}
`;

exports[`cursor reducer should handle EDITOR_MOVE_CURSOR down 2`] = `
Object {
  "charIndex": 0,
  "endOffset": 0,
  "lineIndex": 1,
  "startOffset": 0,
}
`;

exports[`cursor reducer should handle EDITOR_MOVE_CURSOR left 1`] = `
Object {
  "charIndex": 0,
  "lineIndex": 0,
}
`;

exports[`cursor reducer should handle EDITOR_MOVE_CURSOR left 2`] = `
Object {
  "charIndex": 2,
  "lineIndex": 0,
}
`;

exports[`cursor reducer should handle EDITOR_MOVE_CURSOR right 1`] = `
Object {
  "charIndex": 0,
  "lineIndex": 0,
}
`;

exports[`cursor reducer should handle EDITOR_MOVE_CURSOR up 1`] = `
Object {
  "charIndex": 0,
  "endOffset": 0,
  "lineIndex": 0,
  "startOffset": 0,
}
`;

exports[`cursor reducer should handle EDITOR_MOVE_CURSOR up 2`] = `
Object {
  "charIndex": 0,
  "endOffset": 0,
  "lineIndex": 0,
  "startOffset": 0,
}
`;

exports[`cursor reducer should handle EDITOR_MOVE_CURSOR up 3`] = `
Object {
  "charIndex": 0,
  "endOffset": 0,
  "lineIndex": 0,
  "startOffset": 0,
}
`;

exports[`cursor reducer should handle EDITOR_MOVE_CURSOR up 4`] = `
Object {
  "charIndex": 0,
  "endOffset": 0,
  "lineIndex": 1,
  "startOffset": 0,
}
`;

exports[`cursor reducer should handle EDITOR_MOVE_CURSOR up 5`] = `
Object {
  "charIndex": 1,
  "endOffset": 1,
  "lineIndex": 1,
  "startOffset": 1,
}
`;

exports[`cursor reducer should return the initial state 1`] = `
Object {
  "charIndex": 0,
  "endOffset": 0,
  "lineIndex": 0,
  "startOffset": 0,
}
`;
